File: c++_code/347.前-k-个高频元素.cpp
/*
 * @lc app=leetcode.cn id=347 lang=cpp
 *
 * [347] 前 K 个高频元素
 */

// @lc code=start
#include <stdio.h>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;
bool cmp(pair<int,int> a, pair<int,int> b) {
	return a.second > b.second;
}
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
       map<int, int> mp;
       for (int i = 0; i < nums.size();i++)
       {
           mp[nums[i]] +=1;
       }
       vector< pair<int,int> > vec;
       for(map<int,int>::iterator it = mp.begin(); it != mp.end(); it++){
           vec.push_back( pair<int,int>(it->first,it->second) );
       }
       sort(vec.begin(),vec.end(),cmp);
       vector<int> ans;
       for(int i=0;i<k;i++)
       {
           ans.push_back(vec[i].first);
       }
       return ans;
    }
};
// @lc code=end

File: js_code/专项/哈希/347. 前 K 个高频元素.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
`
给你一个整数数组 nums 和一个整数 k ，请你返回其中出现频率前 k 高的元素。你可以按 任意顺序 返回答案。
`
`
1.首先新建一个Map对象，然后使用循环对元素的频率进行统计
2.接着将字典转成数组，然后对数组中的第二个元素（频度）从小到大排序
3.最后截取频率前k高的元素返回即可
`
 var topKFrequent = function(nums, k) {

  let map=new Map()
  for(let n of nums){
      map.set(n,map.has(n)?map.get(n)+1:1)
  }

  let ans=Array.from(map).sort((a,b)=>b[1]-a[1]) //按照键值从大到小排序
  return ans.slice(0,k).map(item=>item[0])

};